src/ucp/proto/proto_multi.c

```diff
@@ -100,25 +98,64 @@ ucs_status_t ucp_proto_multi_init(const ucp_proto_multi_init_params_t *params,
     mpriv->lane_map   = lane_map;
     mpriv->num_lanes  = 0;
     perf.max_frag     = SIZE_MAX;
+    perf.min_length   = 0;
     ucs_for_each_bit(lane, lane_map) {
         ucs_assert(lane < UCP_MAX_LANES);
 
         lpriv     = &mpriv->lanes[mpriv->num_lanes++];
         lane_perf = &lanes_perf[lane];
-        max_frag  = ucs_double_to_sizet(lane_perf->bandwidth / max_frag_ratio,
-                                        lane_perf->max_frag);
 
         ucp_proto_common_lane_priv_init(&params->super, mpriv->reg_md_map, lane,
                                         &lpriv->super);
 
-        /* Due to floating-point accuracy, max_frag may be too large */
+        /* Calculate maximal fragment size according to lane relative bandwidth.
+           Due to floating-point accuracy, max_frag may be too large. */
+        max_frag = ucs_double_to_sizet(lane_perf->bandwidth / max_frag_ratio,
+                                       lane_perf->max_frag);
+        ucs_assert(max_frag > 0);
+
         lpriv->max_frag = max_frag;
-        lpriv->weight   = ucs_proto_multi_calc_weight(lane_perf->bandwidth,
-                                                      perf.bandwidth);
         perf.max_frag   = ucs_min(perf.max_frag, lpriv->max_frag);
 
-        ucs_assert(lpriv->max_frag > 0);
+
+        /* Calculate lane weight as a fixed-point fraction */
+        lpriv->weight = ucs_proto_multi_calc_weight(lane_perf->bandwidth,
+                                                    perf.bandwidth);
         ucs_assert(lpriv->weight > 0);
+
+        /* Calculate minimal message length according to lane's relative weight:
+           When the message length is scaled by this lane's weight, it must not
+           be lower than 'lane_perf->min_length'. Proof of the below formula:
+
+            Since we calculated min_length by our formula:
+                length >= (lane_perf->min_length << SHIFT) / weight
+
+            Let's mark "(lane_perf->min_length << SHIFT)" as "M" for simplicity:
+                length >= M / weight
+
+            Since weight <= mask + 1, then weight - (mask + 1) <= 0, so we get:
+                length >= (M + weight - (mask + 1)) / weight)
+
+            Changing order:
+                length >= ((M - mask) + weight - 1) / weight)
+
+            The righthand expression is actually rounding up the result of
+            dividing the "M - mask" by weight, so we get:
```

Comment:
maybe add more explanations why the righthand expressions can be shortened: 
`... dividing the "M - mask" by "weight", since dividing "weight - 1" by "weight" > 0, so we get:`